onnxruntime/contrib_ops/cpu/signal/dft.cc

```diff
@@ -0,0 +1,521 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#ifdef BUILD_MS_EXPERIMENTAL_OPS
+
+#include "core/providers/common.h"
+#include "core/framework/op_kernel.h"
+#include "core/util/math_cpuonly.h"
+#include "Eigen/src/Core/Map.h"
+#include "dft.h"
+#include <functional>
+
+#include "core/platform/threadpool.h"
+
+#include <complex>
+
+namespace onnxruntime {
+namespace contrib {
+
+ONNX_OPERATOR_KERNEL_EX(
+    DFT,
+    kMSExperimentalDomain,
+    1,
+    kCpuExecutionProvider,
+    KernelDefBuilder().MayInplace(0, 0).TypeConstraint("T", BuildKernelDefConstraints<float, double>()),
+    DFT);
+
+ONNX_OPERATOR_KERNEL_EX(
+    IDFT,
+    kMSExperimentalDomain,
+    1,
+    kCpuExecutionProvider,
+    KernelDefBuilder().MayInplace(0, 0).TypeConstraint("T", BuildKernelDefConstraints<float, double>()),
+    IDFT);
+
+ONNX_OPERATOR_KERNEL_EX(
+    STFT,
+    kMSExperimentalDomain,
+    1,
+    kCpuExecutionProvider,
+    KernelDefBuilder().MayInplace(0, 0).TypeConstraint("T", BuildKernelDefConstraints<float, double>()),
+    STFT);
+
+static bool is_real_valued_signal(const onnxruntime::TensorShape & shape) {
+  // The first dimention is the batch size
+  // The second dimention is the signal value
+  return shape.NumDimensions() == 2;
+}
+
+static bool is_complex_valued_signal(const onnxruntime::TensorShape& shape) {
+  // The first dimention is the batch size
+  // The second dimention is the signal length
+  // The third dimention is set to 2 and represents the real and imaginary parts of the complex sample
+  return shape.NumDimensions() == 3 && shape[2] == 2;
+}
+
+static bool is_power_of_2(size_t size) {
+  unsigned n_bits = 0;
+  while (size != 0) {
+    n_bits += size & 1;
+    size = size >> 1;
+  }
+  return n_bits == 1;
+}
```

Comment:
You can do it with some simple binary operations: https://stackoverflow.com/questions/600293/how-to-check-if-a-number-is-a-power-of-2

Reply:
I'll probably redo it with the simpler implementation after this large change goes in.